File: problems/0006_zigzag-conversion/Solution.java
class Solution {
    public String convert(String s, int numRows) {
        final int sLen = s.length();
        if (sLen <= 2 || numRows == 1)
            return s;
        final int colsPerPad = numRows - 1,
                  charsPerPad = colsPerPad - 1 + numRows,
                  lastRow = colsPerPad;
        final StringBuilder buf = new StringBuilder(sLen);
        for (int r = 0; r < numRows; ++r)
            for (int i = r; i < sLen; i += charsPerPad) {
                buf.append(s.charAt(i));
                if (r != 0 && r != lastRow) {
                    final int j = i + charsPerPad - 2 * r;
                    if (j < sLen)
                        buf.append(s.charAt(j));
                }
            }
        return buf.toString();
    }
}

File: problems/0006_zigzag-conversion/Solution.kt
class Solution {
    fun convert(s: String, numRows: Int): String {
        if (s.length <= 2 || numRows == 1)
            return s
        val colsPerPad = numRows - 1
        val charsPerPad = colsPerPad - 1 + numRows
        val lastRow = colsPerPad
        val buf = StringBuilder(s.length)
        for (r in 0 .. (numRows - 1))
            for (i in r .. (s.length - 1) step charsPerPad) {
                buf.append(s[i])
                if (r != 0 && r != lastRow) {
                    val j = i + charsPerPad - 2 * r;
                    if (j < s.length)
                        buf.append(s[j])
                }
            }
        return buf.toString()
    }
}
